modules/aws-http4s/src/smithy4s/aws/AwsClient.scala

```diff
@@ -0,0 +1,101 @@
+/*
+ *  Copyright 2021-2022 Disney Streaming
+ *
+ *  Licensed under the Tomorrow Open Source Technology License, Version 1.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *     https://disneystreaming.github.io/TOST-1.0.txt
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package smithy4s.aws
+
+import cats.effect.Async
+import cats.effect.Resource
+import cats.syntax.all._
+import smithy4s.http4s.kernel._
+import smithy4s.aws.internals._
+import _root_.aws.api.{Service => AwsService}
+
+object AwsClient {
+
+  def apply[Alg[_[_, _, _, _, _]], F[_]: Async](
+      service: smithy4s.Service[Alg],
+      awsEnv: AwsEnvironment[F]
+  ): Resource[F, service.Impl[F]] =
+    prepare(service)
+      .leftWiden[Throwable]
+      .map(_.build(awsEnv))
+      .liftTo[Resource[F, *]]
+
+  def prepare[Alg[_[_, _, _, _, _]]](
+      service: smithy4s.Service[Alg]
+  ): Either[AwsClientInitialisationError, AWSInterpreterBuilder[Alg]] =
+    for {
+      awsService <- service.hints
+        .get(AwsService)
+        .toRight(AwsClientInitialisationError.NotAws(service.id))
+      awsProtocol <- AwsProtocol(service.hints).toRight(
+        AwsClientInitialisationError.UnsupportedProtocol(
+          serviceId = service.id,
+          knownProtocols = AwsProtocol.supportedProtocols
+        )
+      )
+    } yield new AWSInterpreterBuilder(awsProtocol, awsService, service)
+
+  final class AWSInterpreterBuilder[Alg[_[_, _, _, _, _]]](
+      awsProtocol: AwsProtocol,
+      awsService: AwsService,
+      val service: smithy4s.Service[Alg]
+  ) {
+
+    private def interpreter[F[_]: Async](
+        awsEnv: AwsEnvironment[F]
+    ): service.FunctorInterpreter[F] = {
+      val clientCodecs: UnaryClientCodecs[F] = awsProtocol match {
```

Comment:
this is where the money-shot is, which entirely justifies everything that 0.18 is about : the difference in protocols is entirely driven by the codecs. The rest of the logic is entirely decoupled from it, be that signature / routing / etc / etc.